verus/lcf/src/kernel.rs

```diff
@@ -0,0 +1,1108 @@
+                        flag <==> forall|j: int|
+                            #![auto]
+                            0 <= j < i ==> args[j].deep_view().spec_concrete(),
+                {
+                    flag = args[i].clone().concrete() && flag;
+                }
+                flag
+            },
+            Prop::Eq(x, y) => x.concrete() && y.concrete(),
+        }
+    }
+
+    //checks if Prop variants contain key in Susbt TmpStringHashMap
+    pub fn complete_subst(&self, s: &Subst) -> (res: bool)
+        ensures
+            res <==> self.deep_view().spec_complete_subst(s.deep_view()),
+    {
+        match self {
+            Prop::App(head, args) => {
+                //assertion uses forall in App case to show that spec and exec versions are equal
+                assert(match self.deep_view() {
+                    SpecProp::App(_, spec_args) => forall|k: int|
+                        0 <= k < args.len() ==> (#[trigger] args[k].deep_view()) == spec_args[k],
+                    SpecProp::Eq(_, _) => true,
+                });
+                //Next block of code uses a flag calculation instead of forall statement
+                let mut flag = true;
+                for i in 0..args.len()
+                    invariant
+                        0 <= i <= args.len(),
+                        //forall statement used in invariant to reason about complete_subst property for all indices of args seq
+                        flag <==> forall|j: int|
+                            #![auto]
+                            0 <= j < i ==> args[j].deep_view().spec_complete_subst(s.deep_view()),
+                {
+                    flag = args[i].clone().complete_subst(s) && flag;
+                }
+                flag
+            },
+            Prop::Eq(x, y) => x.clone().complete_subst(s) && y.clone().complete_subst(s),
+        }
+    }
+
+    //performs substitution on Prop types and converts them to base types
+    pub fn subst(&self, s: &Subst) -> (res: Prop)
+        requires
+            self.deep_view().spec_complete_subst(s.deep_view()),
+        ensures
+            res.deep_view() == self.deep_view().spec_subst(s.deep_view()),
+            res.deep_view().spec_concrete(),
+    {
+        match self {
+            Prop::App(h, args) => {
+                //assertion uses forall in App case to show that spec and exec versions are equal
+                assert(match self.deep_view() {
+                    SpecProp::App(_, spec_args) => forall|k: int|
+                        0 <= k < args.len() ==> (#[trigger] args[k].deep_view()) == spec_args[k],
+                    SpecProp::Eq(_, _) => true,
+                });
+                let mut v = Vec::<Term>::new();
+                for i in 0..args.len()
+                //loop invariants reason about properties such as complete_subst, concrete, and the result of subst
```

Comment:
The auto-formatter is being a bit unhelpful here---I'd just remove the comments about the invariants, or move them before the `for` loop